docs/src/examples/HyperparameterOptimization.md

```diff
@@ -0,0 +1,425 @@
+
+## General definitions
+
+Here are some general definitions that you will most likely be able to directly use for your problem without any changes.
+Just remember to install `optuna`, for example using `CondaPkg` Julia library.
+
+``` julia
+using Manifolds, Manopt
+using PythonCall
+using BenchmarkTools
+using LineSearches
+
+# This script requires optuna to be available through PythonCall
+# You can install it for example using
+# using CondaPkg
+# ]conda add optuna
+
+optuna = pyimport("optuna")
+
+norm_inf(M::AbstractManifold, p, X) = norm(X, Inf)
+
+struct TTsuggest_int
+    suggestions::Dict{String,Int}
+end
+function (s::TTsuggest_int)(name::String, a, b)
+    return s.suggestions[name]
+end
+struct TTsuggest_float
+    suggestions::Dict{String,Float64}
+end
+function (s::TTsuggest_float)(name::String, a, b; log::Bool=false)
+    return s.suggestions[name]
+end
+struct TTsuggest_categorical
+    suggestions::Dict{String,Any}
+end
+function (s::TTsuggest_categorical)(name::String, vals)
+    return s.suggestions[name]
+end
+struct TTreport
+    reported_vals::Vector{Float64}
+end
+function (r::TTreport)(val, i)
+    return push!(r.reported_vals, val)
+end
+struct TTshould_prune end
+(::TTshould_prune)() = Py(false)
+struct TracingTrial
+    suggest_int::TTsuggest_int
+    suggest_float::TTsuggest_float
+    suggest_categorical::TTsuggest_categorical
+    report::TTreport
+    should_prune::TTshould_prune
+end
+```
+
+        CondaPkg Found dependencies: /home/mateusz/.julia/packages/PythonCall/wXfah/CondaPkg.toml
+        CondaPkg Found dependencies: /home/mateusz/.julia/environments/v1.10/CondaPkg.toml
+        CondaPkg Dependencies already up to date
+
+The next part is your hyperparameter optimization objective. The `ObjectiveData` struct contains all relevant information about the sequence of specific problems.
+The outermost key part is the `N_range` field.
+Early stopping requires a series of progressively more complex problems.
+They will be attempted from the most simple one to the most complex one, and are specified by the values of `N` in that vector.
+
+``` julia
+mutable struct ObjectiveData{TObj,TGrad}
+    obj::TObj
+    grad::TGrad
+    N_range::Vector{Int}
+    gtol::Float64
+    vts::Vector{AbstractVectorTransportMethod}
+    retrs::Vector{AbstractRetractionMethod}
+    manifold_constructors::Vector{Tuple{String,Any}}
+    pruning_losses::Vector{Float64}
+    manopt_stepsize::Vector{Tuple{String,Any}}
+    obj_loss_coeff::Float64
+end
+
+
+function compute_pruning_losses(
+    od::ObjectiveData,
+    int_suggestions::Dict{String,Int},
+    float_suggestions::Dict{String,Float64},
+    categorical_suggestions::Dict{String,Int},
+)
+    tt = TracingTrial(
+        TTsuggest_int(int_suggestions),
+        TTsuggest_float(float_suggestions),
+        TTsuggest_categorical(categorical_suggestions),
+        TTreport(Float64[]),
+        TTshould_prune(),
+    )
+    od(tt)
```

Comment:
This line I do not understand – `od` is just a struct from the current definitions with quite some fields, how can it be called with one parameter?

Reply:
It has a method written later. Anyway, `compute_pruning_losses` is another technical detail, I will move it to another cell.

Reply:
Add at least a comment what that does here, or move it to later (yeah saw it later as well then).